### PR TITLE
[Backend] TF backend

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by yzh119** [comment]
It seems you haven't implemented the logic of BinaryReduce and CopyReduce, right?
Do TF have similar APIs as autograd in PyTorch/MXNet?

**Comment by VoVAllen** [comment]
Yes. There's API called tf.custom_gradient. But still investigating how to store intermediate caches with tf interface.

**Comment by VoVAllen** [comment]
试了一下比想象中简单很多……

**Comment by VoVAllen** [comment]
@jermainewang Tensorflow doesn't support inplace update on tensors. Therefore `scatter_row_inplace` won't be applicable to tf. Could you take a look on our existing implementation to see how could we fix this?

**Comment by jermainewang** [comment]
Let's skip this issue for now by just using outplace update. You may leave
a TODO note there.

**Comment by VoVAllen** [comment]
https://github.com/dmlc/dgl/blob/e2b7f745cac29802d596b607ee99058ef8ce17e2/python/dgl/utils.py#L348
https://github.com/dmlc/dgl/blob/9b4d60799a25409506a73ef4fbaa640a9c4e9ee2/python/dgl/runtime/scheduler.py#L1019

The usage above prevents tf using message passing on the partial graph. @jermainewang

**Comment by jermainewang** [comment]
What's the semantics of `X[index] = value` in tensorflow? Is it inplace or not?

**Comment by VoVAllen** [comment]
Raise error. 
image

**Comment by jermainewang** [comment]
> https://github.com/dmlc/dgl/blob/e2b7f745cac29802d596b607ee99058ef8ce17e2/python/dgl/utils.py#L348
> 
> 
> https://github.com/dmlc/dgl/blob/9b4d60799a25409506a73ef4fbaa640a9c4e9ee2/python/dgl/runtime/scheduler.py#L1019
> 
> The usage above prevents tf using message passing on the partial graph. @jermainewang

How about changing it to use out-place scatter?

**Comment by VoVAllen** [comment]
Yes, but I'm not sure about the interfaces. I'll patch it using out-place scatter at my side for now.

**Comment by VoVAllen** [comment]
Inplcae operation: https://github.com/dmlc/dgl/blob/419c713fd88261d67b02f0cdde0b5c751b2b9da7/python/dgl/nodeflow.py#L1045

**Comment by VoVAllen** [comment]
This is ready for review. @jermainewang

**Comment by VoVAllen** [comment]
| GCN MP | Pytorch GCN MP | GCN builtin | Pytorch GCN builtin
-- | -- | -- | -- | --
CPU (s/epoch)| 0.215 | 0.12 | 0.0467 | 0.0301
GPU (s/epoch)| 0.33 | 0.048 | 0.0216 | 0.0038

**Comment by futurely** [comment]
What are the differences among the last four columns in the first row? Why is GCN MP with GPU slower than GCN MP with CPU?

**Comment by VoVAllen** [comment]
@futurely We are still investigating this issue. Please stay tuned.

**Comment by VoVAllen** [comment]
With default config on cora

  | GCN MP | Pytorch GCN MP | GCN builtin | Pytorch GCN builtin
-- | -- | -- | -- | --
CPU (s/epoch)| 0.213 | 0.12 | 0.0467 | 0.0301
GPU (s/epoch)| 0.310 | 0.048 | 0.0158 | 0.0038

**Comment by VoVAllen** [comment]
I updated to the tf-gpu-nightly version. Seems faster than before

  | GCN MP | Pytorch GCN MP | GCN builtin | Pytorch GCN builtin
-- | -- | -- | -- | --
CPU (s/epoch)| 0.177 | 0.12 | 0.0429 | 0.0301
GPU (s/epoch)| 0.272 | 0.048 | 0.0158 | 0.0038

**Comment by VoVAllen** [comment]
https://github.com/dmlc/dgl/blob/e2b7f745cac29802d596b607ee99058ef8ce17e2/python/dgl/utils.py#L348
https://github.com/dmlc/dgl/blob/9b4d60799a25409506a73ef4fbaa640a9c4e9ee2/python/dgl/runtime/scheduler.py#L1019
https://github.com/dmlc/dgl/blob/419c713fd88261d67b02f0cdde0b5c751b2b9da7/python/dgl/nodeflow.py#L1045

I haven't change the above codes. I'm thinking of change the inplace operation such as `F.scatter_row_inplace(old_to_new, x, F.arange(0, len(x)))` to `old_to_new = F.scatter_row_inplace(old_to_new, x, F.arange(0, len(x)))` and fallback to non_inplace operation when it's not supported. @jermainewang How do you think about this?

**Comment by VoVAllen** [comment]
PyTorch GCN message passing 
inplace vs outplace
CPU: 0.122 vs 0.124
GPU: 0.0475 vs 0.0497

File: docker/Dockerfile.ci_cpu [hunk 1]
```diff
@@ -10,13 +10,6 @@ RUN bash /install/ubuntu_install_core.sh
 COPY install/ubuntu_install_build.sh /install/ubuntu_install_build.sh
 RUN bash /install/ubuntu_install_build.sh
 
-# ANTLR deps
-COPY install/ubuntu_install_java.sh /install/ubuntu_install_java.sh
-RUN bash /install/ubuntu_install_java.sh
-
-COPY install/ubuntu_install_antlr.sh /install/ubuntu_install_antlr.sh
-RUN bash /install/ubuntu_install_antlr.sh
-
```
Comment (jermainewang) [inline docker/Dockerfile.ci_cpu]: Why delete? I think this is required by Jenkins?

Comment (VoVAllen) [reply docker/Dockerfile.ci_cpu]: Why jenkins need this? ANTLR is a AST analysis tool? 
I'll remove all ci related change in this PR and put it at https://github.com/dmlc/dgl/pull/1054

File: examples/tensorflow/gcn/gcn_builtin.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    def call(self, features):
+        h = features
+        for layer in self.layers:
+            h = layer(h)
+        return h
+
+
+def evaluate(model, features, labels, mask):
+    # model.eval()
+    # with torch.no_grad():
```
Comment (jermainewang) [inline examples/tensorflow/gcn/gcn_builtin.py]: remove

File: examples/tensorflow/gcn/gcn_mp.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    def call(self, features):
+        h = features
+        for layer in self.layers:
+            h = layer(h)
+        return h
+
+
+def evaluate(model, features, labels, mask):
+    # model.eval()
+    # with torch.no_grad():
```
Comment (jermainewang) [inline examples/tensorflow/gcn/gcn_mp.py]: Remove

**Comment by jermainewang** [comment]
Could you add a REAME for each example?

File: python/dgl/backend/tensorflow/tensor.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+import tfdlpack
+import numpy as np
+from tfdlpack import to_dlpack, from_dlpack
+
+from ..._ffi.object import ObjectBase
+from ... import ndarray as nd
+from ... import kernel as K
+# import dgl.graph_index
+# from ... import GraphIndex
+# from dgl.graph_index import GraphIndex
```
Comment (jermainewang) [inline python/dgl/backend/tensorflow/tensor.py]: Remove

File: python/dgl/backend/tensorflow/tensor.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+
+
+def repeat(input, repeats, dim):
+    tile_shape = np.ones(input.ndim, dtype=np.int32)
+    tile_shape[dim] = repeats
+    return tf.tile(input, tf.constant(tile_shape))
+
+
+def gather_row(data, row_index):
+    # Weird implementation, because tf.gather require one dimension to be batch_dim
```
Comment (jermainewang) [inline python/dgl/backend/tensorflow/tensor.py]: What is the specific problem here?

File: python/dgl/backend/tensorflow/tensor.py [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+    return tf.gather(data, row_index)
+
+
+def slice_axis(data, axis, begin, end):
+    # assert axis == 0
+    # tf doesn't behave well with negative
```
Comment (jermainewang) [inline python/dgl/backend/tensorflow/tensor.py]: With negative what? And shall we assert the input to be nonnegative?

File: python/dgl/backend/tensorflow/tensor.py [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+def unsorted_1d_segment_sum(input, seg_id, n_segs, dim):
+    assert dim == 0  # Why we need dim for 1d?
+    return tf.math.unsorted_segment_sum(input, seg_id, n_segs)
+
+
+def unsorted_1d_segment_mean(input, seg_id, n_segs, dim):
+    assert dim == 0  # Why we need dim for 1d?
+    return tf.math.unsorted_segment_mean(input, seg_id, n_segs)
+
+# TODO: TF has unsorted_segment_max, which can accelerate _max_on on batched graph
```
Comment (jermainewang) [inline python/dgl/backend/tensorflow/tensor.py]: @yzh119 Could you help review the four functions above? They are related to the glob module you wrote.

File: python/dgl/backend/tensorflow/tensor.py [hunk 8]
```diff
@@ -0,0 +1,564 @@
...
+def nonzero_1d(input):
+    nonzero_bool = (input != False)
+    return tf.reshape(tf.where(nonzero_bool), (-1, ))
+
+
+def sort_1d(input):
+    return tf.sort(input), tf.cast(tf.argsort(input), dtype=tf.int64)
+
+
+def arange(start, stop, ctx=None):
```
Comment (jermainewang) [inline python/dgl/backend/tensorflow/tensor.py]: The backend interface does not require the ctx argument. Could you double check?

File: python/dgl/backend/tensorflow/tensor.py [hunk 9]
```diff
@@ -0,0 +1,564 @@
...
+def rand_shuffle(arr):
+    return tf.random.shuffle(arr)
+
+
+def zerocopy_to_dlpack(input):
+    return tfdlpack.to_dlpack(input)
+
+
+def zerocopy_from_dlpack(dlpack_tensor):
+    # Not zero copy
```
Comment (jermainewang) [inline python/dgl/backend/tensorflow/tensor.py]: Is it?

File: python/dgl/backend/tensorflow/tensor.py [hunk 10]
```diff
@@ -0,0 +1,564 @@
...
+    return nd.from_dlpack(zerocopy_to_dlpack(input))
+
+
+def zerocopy_from_dgl_ndarray(input):
+    return zerocopy_from_dlpack(input.to_dlpack())
+
+
+def one_hot(t, num_classes=-1):
+    raise NotImplementedError
+    # return th.nn.functional.one_hot(t, num_classes)
```
Comment (jermainewang) [inline python/dgl/backend/tensorflow/tensor.py]: Is it used? If not, we could delete it from the backend interface.

File: python/dgl/backend/tensorflow/tensor.py (hunk 5 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: No problem here. I wrote this when I was not familiar with tf API.

File: python/dgl/backend/tensorflow/tensor.py (hunk 6 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: I thinks so. For example end cannot be -1 in tf.

File: python/dgl/backend/tensorflow/tensor.py (hunk 9 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: Yes....I forgot to clean the comment here.

File: python/dgl/backend/tensorflow/tensor.py (hunk 10 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: Sure

File: python/dgl/backend/tensorflow/tensor.py (hunk 5 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: No problem here. I wrote this when I was not familiar with tf API.

File: tests/backend/tensorflow/__init__.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+
+def attach_grad(x):
+    cgrad.add_tensor(x)
+    return x
+    # if x.grad is not None:
+    #     x.grad.zero_()
+    #     return x
+    # else:
+    #     return x.requires_grad_()
```
Comment (jermainewang) [inline tests/backend/tensorflow/__init__.py]: remove

File: tests/backend/tensorflow/__init__.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+
+
+def backward(x, head_gradient=None):
+    cgrad.backward(x, head_gradient)
+    # x.backward(head_gradient)
```
Comment (jermainewang) [inline tests/backend/tensorflow/__init__.py]: same

File: tests/backend/tensorflow/__init__.py [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+
+def grad(x):
+    return cgrad.grad(x)
+    # return x.grad
```
Comment (jermainewang) [inline tests/backend/tensorflow/__init__.py]: same for `grad`, `is_no_grad` functions

**Comment by VoVAllen** [comment]
@zheng-da Since Tensorflow doesn't support inplace update, I changed the code below to`inplace=False`. Do you have any suggestions?
https://github.com/dmlc/dgl/pull/978/files#diff-96cdb6dc3aca65509cabe238780f025fR1045

File: CMakeLists.txt [hunk 14]
```diff
@@ -25,6 +25,8 @@ endif()
 dgl_option(USE_CUDA "Build with CUDA" OFF)
 dgl_option(USE_OPENMP "Build with OpenMP" ON)
 dgl_option(BUILD_CPP_TEST "Build cpp unittest executables" OFF)
+# Set debug compile option for gdb, only happens when -DCMAKE_BUILD_TYPE=DEBUG
+set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O0 -g3 -ggdb")
```
[the remark this answers is not in the document]
Comment (VoVAllen) [reply CMakeLists.txt]: Added if statement. Only add this when not MSVC. Checked clang supported this.

File: python/dgl/backend/tensorflow/tensor.py (hunk 6 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: I made a work around for negative index in this implemetation

File: python/dgl/backend/tensorflow/tensor.py (hunk 8 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: Fixed

File: python/dgl/backend/tensorflow/tensor.py (hunk 10 above)
Comment (VoVAllen) [reply python/dgl/backend/tensorflow/tensor.py]: Removed

**Comment by VoVAllen** [comment]
image
终终终终终终终终终终终终终终终终终终终终终终终终终终终终于搞好了

@jermainewang Ready for merge

**Comment by curiosityyy** [comment]
> image
> 终终终终终终终终终终终终终终终终终终终终终终终终终终终终于搞好了
> 
> @jermainewang Ready for merge. KG now uses same image as others.

你好，我想问一下，那个segmentation fault 具体是什么问题，这里哪里的代码是fix那个问题的。

**Comment by VoVAllen** [comment]
@curiosityyy 
https://github.com/dmlc/dgl/pull/978/files#diff-06c69bd0cdfc13afd730c7658460da97

We temporarily removed deconstruct of workspacepool, since cuda library will wipe out used memorys. However this seems a bug with mxnet, we are still trying to figure out why.

**Comment by zhangxin9988** [comment]
hello,I encounted a problem.don't know how to solov it.
I want to use dgl-cpu version,  and I installed tensorflow=2.0.0,tfdlpack-0.1.2 on my Ubuntu18.04
and I want to export tensorflow as backend
below is the code  and the error:
import os
os.environ['DGLBACKEND']='tensorflow'
import dgl

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/dgl/__init__.py", line 8, in <module>
    from .backend import load_backend
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/dgl/backend/__init__.py", line 48, in <module>
    load_backend(os.environ.get('DGLBACKEND', 'pytorch').lower())
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/dgl/backend/__init__.py", line 18, in load_backend
    mod = importlib.import_module('.%s' % mod_name, __name__)
  File "/home/zngxin/miniconda3/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/dgl/backend/tensorflow/__init__.py", line 1, in <module>
    from .tensor import *
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/dgl/backend/tensorflow/tensor.py", line 9, in <module>
    import tfdlpack
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/tfdlpack/__init__.py", line 2, in <module>
    from .core import __version__, to_dlpack, from_dlpack, get_device_and_dtype
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/tfdlpack/core.py", line 13, in <module>
    dlpack_ops = load_library.load_op_library(libinfo.find_lib_path()[0])
  File "/home/zngxin/miniconda3/lib/python3.7/site-packages/tensorflow_core/python/framework/load_library.py", line 61, in load_op_library
    lib_handle = py_tf.TF_LoadLibrary(library_filename)
tensorflow.python.framework.errors_impl.NotFoundError: /home/zngxin/miniconda3/lib/python3.7/site-packages/tfdlpack/libtfdlpack.so: undefined symbol: _ZN10tensorflow12OpDefBuilder4AttrESs

**Comment by VoVAllen** [comment]
@zhangxin9988 This may due to your tensorflow is installed by `conda install tensorflow` instead of `pip install tensorflow`. There's minor difference there which prevents the tfdlpack to find the corresponding symbol. However, dlpack is supported in the latest tensorflow now. And we plan to release the official support of tensorflow next week. Please stay tuned!

**Pull request merged**